File: solutions/[50]Pow(x, n).cpp
class Solution {
public:
	double binary_pow(double x, int n) {
		if (n == 0)
			return 1.0;
		if (n == 1)
			return x;
		if (n % 2 == 0)
			return binary_pow(x * x, n / 2);
		else
			return binary_pow(x * x, n / 2) * x;
	}
	double myPow(double x, int n) {
		long e = n;
		if (e < 0) {
			x = 1 / x;
			e = -e;
		}

		return binary_pow(x, e);
	}
};

File: solutions/[50]Pow(x,n).py
class Solution:
    def myPow(self, x: float, n: int) -> float:
        if n == 0:
            return 1
        elif n == 1:
            return x
        
        if n < 0:
            x = 1 / x
            n = -n
        return self.myPow(x, n // 2) ** 2 * self.myPow(x, n % 2)
